### PR TITLE
Avoid MemoryMarshal.Cast when transcoding from UTF-16 to UTF-8 while escaping in Utf8JsonWriter.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/40979 in master.

This is meant to be a targeted fix to be ported to 3.0.

cc @steveharter, @GrabYourPitchforks, @pranavkm, @ericstj

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs [hunk 1]
```diff
@@ -79,13 +79,16 @@ public static int NeedsEscaping(ReadOnlySpan<byte> value, JavaScriptEncoder enco
...
             int idx;
 
             if (encoder != null)
             {
-                idx = encoder.FindFirstCharacterToEncodeUtf8(MemoryMarshal.Cast<char, byte>(value));
+                fixed (char* ptr = value)
+                {
+                    idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
+                }
                 goto Return;
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs]: For v5 we may want to add a overload of `FindFirstCharacterToEncode(ReadOnlySpan<char)` to S.T.Encoding.Web so consumers don't have to use unsafe pinning code.

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs [hunk 2]
```diff
@@ -79,13 +79,16 @@ public static int NeedsEscaping(ReadOnlySpan<byte> value, JavaScriptEncoder enco
...
+        public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
             int idx;
 
             if (encoder != null)
             {
-                idx = encoder.FindFirstCharacterToEncodeUtf8(MemoryMarshal.Cast<char, byte>(value));
+                fixed (char* ptr = value)
+                {
+                    idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
```
Comment (GrabYourPitchforks) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs]: Some implementations of the `FindFirstCharacterToEncode` method may not accept null pointers. We should special-case `value.IsEmpty` at the beginning of this method and bail early.

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs [hunk 3]
```diff
@@ -56,6 +56,54 @@ public void CantWriteToNonWritableStream(bool formatted, bool skipValidation)
...
 
+        [Fact]
+        public static void WritingStringsWithCustomEscaping()
+        {
+            var output = new ArrayBufferWriter<byte>();
+            var writerOptions = new JsonWriterOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
+
+            using (var writer = new Utf8JsonWriter(output))
+            {
+                writer.WriteStringValue("\u6D4B\u8A6611");
```
Comment (GrabYourPitchforks) [inline src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: `U+6D4B` and `U+8A66` might be allowed through unescaped in a future version, which could break this unit test. If you're looking for something stable that's highly unlikely to ever be allowed through unescaped, consider something from the range `U+E000..U+F8FF` (inclusive). That block is permanently reserved for private use and I highly doubt even the "relaxed" escaper will ever allow those to pass through unescaped.

If the test doesn't rely on this being output escaped or unescaped, you're good to go. :)

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: > might be allowed through unescaped in a future version

We have a few writer tests that ensure the default behavior escapes certain characters. If we change that behavior, we would have to/should change those tests as well, so I would prefer the tests break at that time.

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs]: Will fix this and add a buggy javascriptencoder implementation as a test.

**Comment by ahsonkhan** [comment]
MacOS Build x64_Debug test failures are **unrelated**:
https://dev.azure.com/dnceng/public/_build/results?buildId=348664&view=ms.vss-test-web.build-test-results-tab&runId=10473744&resultId=168822&paneView=debug

System.Security.Cryptography.OpenSsl.Tests on netcoreapp-OSX-Debug-x64-OSX.1014.Amd64.Open

System.Security.Cryptography.OpenSsl.Tests  Total: 649, Errors: 0, Failed: 565, Skipped: 14, Time: 1.083s

https://github.com/dotnet/corefx/issues/40262
https://github.com/dotnet/corefx/issues/38498

Example test failure:
```text
System.Security.Cryptography.Dsa.Tests.DSAKeyFileTests.ReadWriteDsa1024EncryptedPkcs8_PasswordBytes [FAIL]
      System.TypeInitializationException : The type initializer for 'Crypto' threw an exception.
      ---- System.TypeInitializationException : The type initializer for 'CryptoInitializer' threw an exception.
      -------- System.DllNotFoundException : Unable to load shared library 'System.Security.Cryptography.Native.OpenSsl' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(libSystem.Security.Cryptography.Native.OpenSsl, 1): image not found
      Stack Trace:
           at Interop.Crypto.DsaKeyCreateByExplicitParameters(SafeDsaHandle& dsa, Byte[] p, Int32 pLength, Byte[] q, Int32 qLength, Byte[] g, Int32 gLength, Byte[] y, Int32 yLength, Byte[] x, Int32 xLength)
        /_/src/Common/src/System/Security/Cryptography/DSAOpenSsl.cs(112,0): at System.Security.Cryptography.DSAOpenSsl.ImportParameters(DSAParameters parameters)
        /_/src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs(349,0): at System.Security.Cryptography.DSA.ImportEncryptedPkcs8PrivateKey(ReadOnlySpan`1 passwordBytes, ReadOnlySpan`1 source, Int32& bytesRead)
        /_/src/Common/src/System/Security/Cryptography/DSAOpenSsl.cs(132,0): at System.Security.Cryptography.DSAOpenSsl.ImportEncryptedPkcs8PrivateKey(ReadOnlySpan`1 passwordBytes, ReadOnlySpan`1 source, Int32& bytesRead)
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSAKeyFileTests.cs(593,0): at System.Security.Cryptography.Dsa.Tests.DSAKeyFileTests.<>c__DisplayClass23_0.<ReadBase64EncryptedPkcs8>b__0(DSA dsa, ReadOnlySpan`1 source, Int32& read)
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSAKeyFileTests.cs(652,0): at System.Security.Cryptography.Dsa.Tests.DSAKeyFileTests.ReadWriteKey(String base64, DSAParameters& expected, ReadKeyAction readAction, Func`2 writeArrayFunc, WriteKeyToSpanFunc writeSpanFunc, Boolean isEncrypted)
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSAKeyFileTests.cs(589,0): at System.Security.Cryptography.Dsa.Tests.DSAKeyFileTests.ReadBase64EncryptedPkcs8(String base64EncPkcs8, Byte[] passwordBytes, PbeParameters pbeParameters, DSAParameters& expected)
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSAKeyFileTests.cs(174,0): at System.Security.Cryptography.Dsa.Tests.DSAKeyFileTests.ReadWriteDsa1024EncryptedPkcs8_PasswordBytes()
        ----- Inner Stack Trace -----
        /_/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Initialization.cs(40,0): at Interop.CryptoInitializer.Initialize()
        /_/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Initialization.cs(19,0): at Interop.Crypto..cctor()
        ----- Inner Stack Trace -----
           at Interop.CryptoInitializer.EnsureOpenSslInitialized()
        /_/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Initialization.cs(27,0): at Interop.CryptoInitializer..cctor()
```

**Pull request merged**